src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,256 @@
+#pragma once
+#include <chrono>
+#include <thread>
+#include <condition_variable>
+#include "rgw_common.h"
+
+
+class RateLimiter {
+  static constexpr int64_t fixed_float_rgw_ratelimit = 1000;
+  // counters are tracked in multiples of fixed_float_rgw_ratelimit
+  struct counters {
+    int64_t ops = 0;
+    int64_t bw = 0;
+  };
+  struct qos_entry {
+    // 2 is the enum size
+    counters read;
+    counters write;
+    ceph::timespan ts;
+    bool first_run = true;
+    std::mutex ts_lock;
+    // Those functions are returning the integer value of the tokens 
+    int64_t read_ops () const
+    {
+      return read.ops / fixed_float_rgw_ratelimit;
+    }
+    int64_t write_ops() const
+    {
+      return write.ops / fixed_float_rgw_ratelimit;
+    }
+    int64_t read_bytes() const
+    {
+      return read.bw / fixed_float_rgw_ratelimit;
+    }
+    int64_t write_bytes() const
+    {
+      return write.bw / fixed_float_rgw_ratelimit;
+    }
+  };
+  static constexpr size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::shared_mutex insert_lock;
+  std::atomic_bool *replacing;
+  std::condition_variable *cv;
+  typedef std::unordered_map<std::string, qos_entry> hash_map;
+  hash_map qos_entries{map_size};
+  static bool is_read_op(const std::string_view method) {
+    if(method == "GET" || method == "HEAD")
+      return true;
+    return false;
+  }
+
+    // find or create an entry, and return its iterator
+  auto& find_or_create(const std::string& key) {
+    std::shared_lock rlock(insert_lock);
+    if(qos_entries.size() > 0.9 * map_size && *replacing == false)
+    {
+        *replacing = true;
+        cv->notify_all();
+    }
+    auto ret = qos_entries.find(key);
+    rlock.unlock();
+    if(ret == qos_entries.end())
+    {
+      std::unique_lock wlock(insert_lock);
+      ret = qos_entries.emplace(std::piecewise_construct,
+                                 std::forward_as_tuple(key),
+                                 std::forward_as_tuple()).first;
+    }
+    return ret->second;
+  }
+
+  bool should_rate_limit_read(qos_entry& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
```

Comment:
do we use "0" or "-1" as unlimited?
if we use zero, we probably need to add the to the help in radosgw-admin
also, not sure we have "-1" there? or why do we use signed long long?

Reply:
0 means unlimited I will add it to the radosgw-admin -h
so ops_limit could not  be negative and bw_limit(info->max_read_bytes) could not be too.
info->max_read_bytes will force us to cast it to int64_t each decrease_bytes() (formerly increase_bw()) call as we do **not** want implicit casting to uint64
we can change ops_limit to be uint64, but I'm not sure if we want to seperate their types.
@cbodley what do you think?

Reply:
right, since `int64_t bytes` can go negative due to burst, i think it makes sense for the ops and limits to be expressed in `int64_t` as well. that avoids signed vs. unsigned comparisons, and the logic can be (mostly) identical for ops and bw calculations

i do have a question related to `0 = unlimited`, though.. if `RGWQoSInfo` defaults everything to 0/unlimited, then what's the point of the extra field `RGWQoSInfo::enabled`?

Reply:
Honestly it took me a minute or two to remember why 😌.
There this enable or disable to handle the global and not global, for example I want that a bucket will have specifically no limit and another one will be as the global settings, I will enable on the bucket the policy and then the rgw will take it and not the global and on the other bucket it will take the global